pkg/agent/controller/agent.go

```diff
@@ -263,35 +384,86 @@ func (a *Controller) updateExportedServiceStatus(name, namespace string, condTyp
 	}
 }
 
-func serviceExportConditionEqual(c1, c2 *lighthousev2a1.ServiceExportCondition) bool {
-	return c1.Type == c2.Type && c1.Status == c2.Status && reflect.DeepEqual(c1.Reason, c2.Reason) &&
-		reflect.DeepEqual(c1.Message, c2.Message)
+func (a *Controller) serviceImportEquivalent(obj1, obj2 *unstructured.Unstructured) bool {
+	return syncer.DefaultResourcesEquivalent(obj1, obj2) &&
+		equality.Semantic.DeepEqual(util.GetNestedField(obj1, "status"), util.GetNestedField(obj2, "status"))
 }
 
-func (a *Controller) newClusterStatus(service *corev1.Service, clusterID string) lighthousev2a1.ClusterStatus {
-	mcsIp := getGlobalIpFromService(service)
-	if mcsIp == "" {
-		mcsIp = service.Spec.ClusterIP
-	}
+func (a *Controller) endpointEquivalent(obj1, obj2 *unstructured.Unstructured) bool {
+	return equality.Semantic.DeepEqual(util.GetNestedField(obj1, "subsets"),
+		util.GetNestedField(obj2, "subsets"))
+}
 
-	return lighthousev2a1.ClusterStatus{
-		Cluster: clusterID,
-		IPs:     []string{mcsIp},
-	}
+func serviceExportConditionEqual(c1, c2 *lighthousev2a1.ServiceExportCondition) bool {
+	return c1.Type == c2.Type && c1.Status == c2.Status && reflect.DeepEqual(c1.Reason, c2.Reason) &&
+		reflect.DeepEqual(c1.Message, c2.Message)
 }
 
 func (a *Controller) newServiceImport(svcExport *lighthousev2a1.ServiceExport) *lighthousev2a1.ServiceImport {
 	return &lighthousev2a1.ServiceImport{
 		ObjectMeta: metav1.ObjectMeta{
-			Name: svcExport.Name + "-" + svcExport.Namespace + "-" + a.clusterID,
+			Name: a.getObjectNameWithClusterId(svcExport.Name, svcExport.Namespace),
 			Annotations: map[string]string{
 				originName:      svcExport.Name,
 				originNamespace: svcExport.Namespace,
 			},
+			Labels: map[string]string{
+				labelSourceName:      svcExport.Name,
+				labelSourceNamespace: svcExport.Namespace,
+				labelSourceCluster:   a.clusterID,
+			},
 		},
 	}
 }
 
+func (a *Controller) getServiceImport(svcExport *lighthousev2a1.ServiceExport) (*lighthousev2a1.ServiceImport, error) {
+	siName := a.getObjectNameWithClusterId(svcExport.Name, svcExport.Namespace)
+	serviceImport, err := a.lighthouseClient.LighthouseV2alpha1().ServiceImports(a.namespace).Get(siName, metav1.GetOptions{})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return serviceImport, nil
+}
+
+func (a *Controller) getIpsForService(service *corev1.Service, siType lighthousev2a1.ServiceImportType) []string {
+	if siType == lighthousev2a1.SuperclusterIP {
+		mcsIp := getGlobalIpFromService(service)
+		if mcsIp == "" {
+			mcsIp = service.Spec.ClusterIP
+		}
+
+		return []string{mcsIp}
+	}
+
+	endpoint, err := a.kubeClientSet.CoreV1().Endpoints(service.Namespace).Get(service.Name, metav1.GetOptions{})
+	if err != nil {
+		klog.V(log.DEBUG).Infof("Endpoints for svc  (%s/%s) not found: %v", service.Namespace, service.Name, err)
```

Comment:
Most likely the error is `IsNotFound` but we shouldn't assume it.

```suggestion
                if !errors.IsNotFound(err) {
		    klog.Errorf("Error obtaining Endpoint for Service  (%s/%s): %v", service.Namespace, service.Name, err)
		}
```

Reply:
Actually we should re-queue if error is something other than `NotFound`

Reply:
Done